docs/_posts/2022-10-31-luau-semantic-subtyping.md

```diff
@@ -0,0 +1,248 @@
+---
+layout: single
+title:  "Semantic Subtyping in Luau"
+author: Alan Jeffrey
+---
+
+Luau is the first programming language to put semantic subtyping in the hands of millions of creators.
+
+## Minimizing false positives
+
+One of the issues with type error reporting in tools like the Script Analysis widget in Roblox Studio is *false positives*. These are warnings that are artifacts of the analysis, and don’t correspond to errors which can occur at runtime. For example, the program
+```lua
+  local x
+  x = nil
+  x = 5
+  print(x + 1)
+```
+reports a type error which cannot happen at runtime, since by the time `x + 1` is executed, `x` is `5`, not `nil`.
+
+False positives are especially poor for onboarding new users. If a type-curious creator switches on typechecking and is immediately faced with a wall of spurious red squiggles, there is a strong incentive to immediately switch it off again.
+
+Inaccuracies in type errors are inevitable, since it is undecidable to decide ahead of time whether a runtime error will be triggered. Type system designers have to choose whether to live with false positives or false negatives. In Luau this is determined by the mode: `strict` mode errs on the side of false positives, and `nonstrict` mode errs on the side of false negatives.
+
+While inaccuracies are inevitable, we try to remove them whenever possible, since they result in spurious errors, and imprecision in type-driven tooling like autocomplete or API documentation.
+
+## Subtyping as a source of false positives
+
+One of the sources of false positives in Luau (and many other similar languages like TypeScript or Flow) is *subtyping*. Subtyping is used whenever a variable is initialized or assigned to: the type system checks that the type of the expression is a subtype of the type of the variable. For example in the program
+```lua
+  local x : number? = nil
+  local y : number = 5
+  y = x -- TypeError: Type 'number?' could not be converted into 'number'
```

Comment:
Yeah I think the example you gave is a case where type errors will happen even with typestates. I think you want
```
  local x : number? = 5
  local y : number = 5
  y = x -- TypeError: Type 'number?' could not be converted into 'number'
  x = y
```
Then when typestates is in, it'll check whether `number <: number` (true) and not `nil <: number` (false), or what we do today `number? <: number` (false)`.

I agree with Ben that there should be a type error from `y = x` where `x = nil` because there is an upper bound `number` for `y`, typestates shouldn't ignore that.

Reply:
👍  still seems like a good error to me.

The concern behind my comment is that I had trouble following this part of the section. "What's the purpose of this example; is it a false positive? No I guess not." I think a word or two about what to expect would help a lot.

Reply:
Okay, evidently this example is confusing. I'll see if I can come up with something clearer.

Reply:
I redid the example, hopefully it's clearer, though it does use the Roblox APIs.

Reply:
I was talking about this example (l.33). But yes, that `CFrame` example looks good!

Reply:
OK, I used the `CFrame` example for both.